### ISSUE TITLE
Appraisal v3

### ISSUE THREAD

**Issue by nickcharlton** [thoughtbot/appraisal#204]
It's been a while since the last release and I know folks are keen for a new one. I'm currently tracking/working on the following, some of which are much easier to fix than others but this issue should communicate where we are.

## New Features

* Support for Ruby 3.2: https://github.com/thoughtbot/appraisal/issues/199
* Fix GitHub Actions/Bundle environment issues: https://github.com/thoughtbot/appraisal/pull/202
* Customising Gemfiles: https://github.com/thoughtbot/appraisal/pull/191
* https://github.com/thoughtbot/appraisal/pull/193

## Remaining

* Replacing Travis (which is completely broken now) with GitHub Actions: https://github.com/thoughtbot/appraisal/compare/nc-use-github-actions?expand=1
* Testing against recent combinations of Bundler to see what we last support
* See which of the deprecation warnings we can fix now, what can wait until later

**Comment by nickcharlton** [comment]
I'm currently stuck on understanding why Ruby 2.6, 2.7 and 3.2 fail. The 2.* series are the same reason, but 3.2 is close but not completely the same.

I'm also thinking of dropping 2.5 as it's so old, but with it working with very little effort I'm okay with keeping it for now.

**Comment by joe-sharp** [comment]
Just a suggestion Nick, I would drop support for Ruby 2.5 and 2.6. According to ruby-lang.org they are both past EOL: https://www.ruby-lang.org/en/downloads/branches/

**Comment by nickcharlton** [comment]
Yeah, I suspect I will end up doing that. It's definitely not worth too much effort, even for a library where I want to make things easier for folks.

I'm more concerned with 3.2, but if the solution for that fixes problems then I'd be happy for a little longer. I suspect 3.1 or whatever in six months would then drop it.

**Comment by unicornzero** [comment]
Hi! My app is already on Ruby 3.2 and I need to use dd-trace-rb for monitoring, but am running into an [issue](https://github.com/DataDog/dd-trace-rb/issues/2534) for which [a fix appears to be blocked](https://github.com/DataDog/dd-trace-rb/pull/2497/files#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55fR12) on needing Appraisal support for Ruby 3.2. 

What kind of timeline might we be looking at for Ruby 3.2 support? 

Thanks!

**Comment by nickcharlton** [comment]
It's hard to put a timeline to it, unfortunately. I'm dipping into when I can find a bit of time/inspiration on what might be causing the tests to fail on that CI branch I started but I can only do that in between a busy client project and management responsibilities.

So er, any thoughts are welcome on that CI branch!

**Comment by nickcharlton** [comment]
I've started looking at the deprecation warnings, with the hope that this either solves our CI problem or indicates what it might be, starting with:

* #207 
* #208

**Comment by Maimer** [comment]
I have been trying to get the tests passing on CI and the more I look at it the more I think it's an issue of what `bundler` version is being tested against and less what Ruby version is being tested against. Unfortunately `bundler` doesn't follow semantic versioning and as of recently there have been [lots of releases](https://rubygems.org/gems/bundler/versions). It almost feels like the test matrix should be both Ruby version and `bundler` version, but it's hard to know which ones to test. Should the `appraisal` gem be updated to support a minimum `bundler` version?

**Comment by nickcharlton** [comment]
Yeah, I think that's exactly the problem we're facing.

I'd like to identify the lowest-possible Bundler release we can target and then start testing from that release onwards. If anyone has a good suggestion, that'd be brilliant.

I think then we should be testing major (we'll need to figure that out ourselves, I guess?) releases of Bundler in another matrix on CI.

**Comment by kyrofa** [comment]
Still using ruby 2.5 over here, FYI :wave: . If it's not too much trouble, I would like to see it continue to be supported.

**Comment by Maimer** [comment]
> Yeah, I think that's exactly the problem we're facing.
> 
> I'd like to identify the lowest-possible Bundler release we can target and then start testing from that release onwards. If anyone has a good suggestion, that'd be brilliant.
> 
> I think then we should be testing major (we'll need to figure that out ourselves, I guess?) releases of Bundler in another matrix on CI.

Personally, I wouldn't start support at `bundler` version `2.2` and go up from there. It's hard to gauge how hard it is going to be to be able to support so many different `bundler` versions. Maybe we just start a matrix with each available version and see what passes and what doesn't.

**Comment by Maimer** [comment]
> Still using ruby 2.5 over here, FYI 👋 . If it's not too much trouble, I would like to see it continue to be supported.

What version of `bundler` are you using with Ruby 2.5?

**Comment by nickcharlton** [comment]
@kyrofa, I think a tool like Appraisal should support as wide a set of Ruby versions as possible so unless it becomes clearly prohibitive to keep supporting, 2.5 will stay.

Of course, Ruby 2.5 is one of the ones which is working fine with Bundler 2.3.10 so so far, that'll stay!

**Comment by kyrofa** [comment]
> What version of bundler are you using with Ruby 2.5?

We like cutting edge technology, so v1.17.3 of course. _cough_

**Comment by joe-sharp** [comment]
@nickcharlton is there any tasks I can try to take off your list? I have some free time this weekend and I would love to help get this release out if there is anything I can do to help! 😄

**Comment by kyrofa** [comment]
Not to speak out of turn, but @joe-sharp if you have some time it sounds like he could use some support on the github actions side of things: https://github.com/thoughtbot/appraisal/pull/206 .

**Comment by Maimer** [comment]
Any reason to use Github Actions over something like CircleCI?

**Comment by kyrofa** [comment]
Any reason to use CircleCI over GitHub Actions? Feels like the in-house solution should be the default unless there's a good reason to use something else.

**Comment by nickcharlton** [comment]
@kyrofa is right — it's GitHub Actions/CI that's the main problem right now.

That said, anything you could do to help with the deprecation warnings would be excellent, too. I think that'd be great to launch a new version with no warnings at all for recent Bundler versions.

Finally, knowing which Bundler versions we should support (for Ruby 2.5 and upwards) would be very helpful too.

**Comment by Maimer** [comment]
> Any reason to use CircleCI over GitHub Actions? Feels like the in-house, integrated solution should be the default unless there's a good reason to use something else.

I find CircleCI to be a bit more comprehensive in terms of what it offers (probably because it's a company solely focused on CI), but I will admit my experience with Github Action has been limited. Also, CircleCI is free for open source projects in case that was concern. However, all that matters is we get a working CI pipeline that tests across all versions of ruby and bundler that need to be tested.

**Comment by Maimer** [comment]
> Sorry, @Maimer, I missed your question.
> 
> I've been having a generally better time with GitHub Actions (although, I dislike the centralisation on a single service). Something I'd seen a lot on Administrate was even though we (thoughtbot) were spending a lot with them, open source builds would still sit in a long queue and worse, fail randomly.
> 
> Anyway, for Appraisal any CI is better than no CI.

We use CircleCI at Appfolio (where I work) and I've had very little issue with it and no queue issues or random failures. Working with it every day is probably why I feel more comfortable with it. I have a branch I have been working on that tests a matrix of Ruby versions and Bundler versions. I have been trying to get it passing against a single Ruby version first as I really don't feel there will be any issues for different Ruby versions, but rather Bundler versions. I have it down to the same test failure for all Bundler versions and haven't had success in figuring out how to get it to pass. It's definitely a configuration setup and not an issue with this gem. You, or anyone else, is welcome to take a look at my changes in this branch and see if you have any ideas: https://github.com/Maimer/appraisal/tree/try_circleci (Changes: https://github.com/thoughtbot/appraisal/compare/main...Maimer:appraisal:try_circleci). As I said before, I don't believe Bundler generally follows semantic versioning, so it's hard to know which versions may have issues. So I just chose the latest version for each patch version available.

```yaml
matrix:
  parameters:
    ruby_version:
      - '3.2.1'
      - '3.1.2'
      - '3.0.5'
      - '2.7.6'
      - '2.6.9'
      - '2.5.9'
    bundler_version:
      - '2.4.7'
      - '2.3.26'
      - '2.2.34'
      - '2.1.4'
      - '1.17.3'
 ```

However, there is definitely a unique test failure for Bundler version `2.4.7` that may require a change in the `appraisal` gem to fix. Here is the latest CI run where you can see the failures: https://app.circleci.com/pipelines/github/Maimer/appraisal/55/workflows/b48d6ca2-d165-407d-a184-6f1d06ac3d07. The main failure is around the tests not being able to find `bundler` when the `--path` option is used as it sets the `path` configuration for `bundler` when that flag is used.

**Comment by javierav** [comment]
Any avance on v3 release?

**Comment by nickcharlton** [comment]
Alas, no.

I've not managed to find any time to go back to investigating the outstanding problems — which is partly been because of some more intense client project time recently and also because I bought a house and that's like having a second job for six months, apparently.

Whilst I'm through with both of those now, which is great for me, I'm wary of setting any expectations otherwise right now. I'll make it obvious on either this issue or others when I'm able to make time though!

**Comment by nickcharlton** [comment]
I've just [released v2.5.0](https://github.com/thoughtbot/appraisal/releases/tag/v2.5.0), which at least ships everything done so far and should fix a lot of the recent bug reports.

**Comment by kyrofa** [comment]
Great idea @nickcharlton, thank you for that.

**Comment by luke-hill** [comment]
Any updates here, we're pinning ourselves on this and some of the functionality regarding the customize gemfiles also isn't working for later rubies

**Comment by nickcharlton** [comment]
Last week I was getting back into it (now that I feel I [just about put out the fire that's been Administrate's asset problems](https://github.com/thoughtbot/administrate/releases/tag/v1.0.0.beta1)), [starting off with trying to get GHA working on an old branch I had to refresh my memory of problems](https://github.com/thoughtbot/appraisal/tree/nc-use-github-actions) and attempting to get my head around some Bundler deprecations and incompatibilities. It looks like I'm going to have more time this week too, which is great. 

My main focus is figuring out what on earth is going on with this code base and changes in Bundler, and then coming up with a plan to handle it better. It's been rough recently.

@luke-hill, [is this the issue you're facing](https://github.com/thoughtbot/appraisal/issues/216), or something different? If it's different, could you open a/some new issues so we can track them?

**Comment by luke-hill** [comment]
Sorry, it's a new issue. I think it's partially related to some things that you may have already tackled. So what I'll do is come back full circle in a couple of weeks and try using the latest versions.

FYI the issues I encountered are in CI runs in cucumber-rails somewhere but I can't spot them. So I think what is best is I wait until you've cut a v3 and then try use it on our matrix.

I think the error is closer to the ERR::ENOENT error on ruby 3.2 if memory serves

**Comment by nickcharlton** [comment]
We've just rolled out a huge bunch of changes, including adding GitHub Actions in for CI.

I'd like to make sure we test out `main` a bunch before cutting a release, but now we're ready to do v3.

For Administrate, I started making up release candidates, and that's probably a good idea here too. But a problem for another day.

**Comment by n-rodriguez** [comment]
Hi there! Any news? Maybe we could fix Hound/Rubocop issue before releasing?

**Comment by nickcharlton** [comment]
I've just published [v3.0.0.rc1](https://github.com/thoughtbot/appraisal/releases/tag/v3.0.0.rc1).

I'd like to make some changes to Hound/Rubocop still, plus maybe adjust the README a bit. I also realise we don't have a CHANGELOG in the file itself, so that's something I think we should do too. But those are fairly minor overall.

If everyone who can could give this a spin and let me know how it goes, that'd be wonderful!

**Comment by pboling** [comment]
This gem is extremely useful for splitting dependencies between each version of ruby so they can each depend on explicit versions of gems that work for that version of Ruby (RubyGems & Bundler).  I am beginning to add appraisals to all my libraries, but I'm having to depend on the older version of Appraisals for my libraries which still support Ruby <= 2.2.

The utility of this gem, for most of my use cases, is inversely related to how much it tries to keep pace with Ruby EOL schedule.

Please keep the old ones until it's at least "not easy" to keep them.

As an aside, @nickcharlton I finished the `eval_gemfile` [PR](https://github.com/thoughtbot/appraisal/pull/248) , if it might be able to get into v3 still.

**Comment by nickcharlton** [comment]
> The utility of this gem, for most of my use cases, is inversely related to how much it tries to keep pace with Ruby EOL schedule.
> 
> Please keep the old ones until it's at least "not easy" to keep them.

I have the same feelings here. I've been struggling a bit to articulate how I think about maintaining support, because it's relevant in a bunch of places. I've not found that keeping support for older Ruby (or Rails in the case of Administrate) is particularly onerous, and the desire to "keep on top of" support windows is just busy work which has unpleasant downstream implications and just gives people a hard time for no good reason.

The trouble I've been having with Appraisal is changes in Bundler, and how to understand where the problem is. When something breaks, I realise that I don't understand Appraisal's codebase all that well and I usually end up giving up fixing something myself (because lack of time) and hoping someone else jumps in (fortunately they do, for which I am extremely grateful).

Thanks for your `eval_gemfile` PR! I'll get to those soon. There's been a lot going on and for various reasons, Appraisal ends up coming up last more often than not 😬

**Comment by pboling** [comment]
In fact the eval_gemfile is one that may need to have a guard for older versions of bundler.

~~I will update the PR to add guards.~~

**Comment by pboling** [comment]
In fact the eval_gemfile is one that may need to have a guard for older versions of bundler.

I will update the PR to add guards.

**Comment by pboling** [comment]
I'm not sure if guards are needed yet, but it will become clear, since there are specs for the feature, and I'll be adding all old Rubies, back to 1.9.3 to CI once #250 is merged (or I can do it on that same PR, but I'd prefer to split it up).  Also please turn off Hound!!!

**Comment by nickcharlton** [comment]
Thanks for all of your hard work so far! I'm out of time again this week, but I'll make some time next week to review everything so far.

Something I've been thinking about which would be extremely useful is a matrix of Ruby and Bundler versions (probably with a column too for Appraisal versions). Perhaps someone has done this already, but it seems like a fairly manual problem to solve. I've been thinking of it like this:

| Ruby | Bundler | Appraisal |
|:--|:--|:--|
| 3.5 | v2.6.4 and up | ??? |
| 3.4 | v2.64 and up | ??? |

I'm taking this from the RubyGems/Bundler releases page: https://github.com/rubygems/rubygems/releases

I'm trying to solve two problems here:

1. Understand what we should be doing for support/testing
2. Try and communicate what we've worked on/expect to work so far

What do you think? Is this something you might be able to help with, perhaps?

**Comment by Maimer** [comment]
> Thanks for all of your hard work so far! I'm out of time again this week, but I'll make some time next week to review everything so far.
> 
> Something I've been thinking about which would be extremely useful is a matrix of Ruby and Bundler versions (probably with a column too for Appraisal versions). Perhaps someone has done this already, but it seems like a fairly manual problem to solve. I've been thinking of it like this:
> Ruby 	Bundler 	Appraisal
> 3.5 	v2.6.4 and up 	???
> 3.4 	v2.64 and up 	???
> 
> I'm taking this from the RubyGems/Bundler releases page: https://github.com/rubygems/rubygems/releases
> 
> I'm trying to solve two problems here:
> 
>     1. Understand what we should be doing for support/testing
> 
>     2. Try and communicate what we've worked on/expect to work so far
> 
> 
> What do you think? Is this something you might be able to help with, perhaps?

I had taken a look at this back in 2023 with this branch: https://github.com/thoughtbot/appraisal/compare/main...Maimer:appraisal:try_circleci but didn't end up getting all the tests to pass. I think one of the main issues that I detailed in this comment: https://github.com/thoughtbot/appraisal/issues/204#issuecomment-1456530703 is that `bundler` does not always follow semantic versioning. So when you say you want to test `2.6.4` and up does that mean every patch version? Or do we want to just stick to the latest patch version of every minor version released? There was 24 releases of `bundler` last year, so the test matrix could get quite large depending on what is decided.

The other issue is that `bundler` is a default gem and can be tricky to remove and set a new default, especially if the new version is lower than the default already installed.

I can take another look at getting this working, but first let's decide on what we want to test first.

**Comment by pboling** [comment]
@nickcharlton I maintain exactly that, I think, and I'll add appraisal. :). If my PR #250 gets merged, and released in v3.0, it'll be v3.0 for every minor version of Ruby. 📦 

Unspoken by you was if you intended it to be a matrix of the _minimum_, or _maximum_, compatible versions of things with each minor version of Ruby.

- FastRuby maintains a matrix that does an imperfect job of cataloging the _minimum_: https://www.fastruby.io/blog/ruby/rails/versions/compatibility-table.html
  - I've used it many times, and found it to rarely be the information I want, which is why I built my version tracking the _maximum_ compatible versions

So I'll wait for that PR #250 first, as it dramatically shifts the matrix for the appraisal column, and then I'll start tracking it here 👇 

- Maximum compatible version matrix: https://dev.to/galtzo/matrix-ruby-gem-bundler-etc-4kk7

**Comment by nickcharlton** [comment]
> Unspoken by you was if you intended it to be a matrix of the minimum, or maximum, compatible versions of things with each minor version of Ruby.

I was thinking "minimum", that's very interesting. Maybe I'm thinking about it wrong, hmm.

The problem is usually changes made by Bundler which means the combination of Ruby language version, Bundler version and Appraisal version don't work together, but one known set does. So if we could know a baseline of what we could reasonably expect to work (and test against those) that'd be very helpful.

**Comment by pboling** [comment]
I've resolved the all of the remaining issues in this thread by hard forking appraisal, and releasing as [appraisal2](https://github.com/appraisal-rb/appraisal2).  Also added important new features, e.g., #248.

CC @nickcharlton Happy to merge back together at some point, but it was a significant drag on my productivity to not have all my work in a released version. I've also added you to the appraisal-rb org, in case it makes sense for you to move work there.
CC @mvz

**Comment by luke-hill** [comment]
Just catching up on here as I last visited around 2 years ago.

I hadn't managed to track anything but I'm using the latest  2.x version (2.5), and it seems some of the "running" issues are now gone. I haven't yet tried to see where the new features landed as honestly at the moment I'm swamped at cucumber with 8/9/10 things higher priority including getting a v11 of cucumber-ruby released.

Just posting here to try keep this alive, it would be good to see appraisal v3 come out as it will inevitably make things easier on cucumber-rails (Which uses appraisal quite heavily)

cc/ @nickcharlton

**Comment by pboling** [comment]
@luke-hill [appraisal2](https://github.com/appraisal-rb/appraisal2) is being actively developed with all fixes from this repo applied and released, and many new features, and all of it fully tested with a far more comprehensive CI.

**Comment by n-rodriguez** [comment]
Yes, but it's sad to see repo/efforts duplication

**Comment by kyrofa** [comment]
Yes, it's sad, and it's a good lesson to all of us:

1. Be grateful for your open source maintainers
2. If you happen to be an open source maintainer, the time will come when for whatever reason you're no longer able to maintain your project. It's hard, but try to recognize when that happens, and bring other maintainers on board.

**Comment by pboling** [comment]
I tried to save this project. And I've invited all the maintainers here to the new project, and 2 of 3 accepted! And I want to merge them back together...

Which is to say, little to no effort has been duplicated. There hasn't been any effort here in quite some time.

**Comment by kyrofa** [comment]
Your efforts are appreciated @pboling 🙏

**Comment by pboling** [comment]
One of the downsides of a project being a corporate project, like this one, is that the corporation can lose interest, fail to hand off the percieved authority to a human developer interested in continuing it, and then no one developer feels empowered to do the needful to keep it alive. Thoughtbot, whoever that is, could add me as a maintainer today, and I'd merge the project back together myself, and release a new unforked appraisal. My projects all depend on this project, so it really is at the root of everything I do, and I can't let it stagnate.

Note that this project hasn't been featured on Thoughtbot's website, where they spotlight their many open source projects, in a long time. It seems to be a forgotten baby.

**Comment by javierav** [comment]
If Thoughtbot is no longer interested in maintaining Appraisal, as seems evident, another option would be to detach it from their organization and transfer it to the community — for example, by creating an "appraisalrb" organization (or similar) on GitHub!

**Comment by pboling** [comment]
That's precisely what I advocated for, and the reason I created https://github.com/appraisal-rb ;) It is the home of `appraisal2`, but I'd be happy to make it the home of the OG `appraisal`.

@nickcharlton - all of the main contributors here are welcome there (as owners). Most have already accepted invites. Would love to make you an owner. LMK your thoughts.
